File: 1-Two Sum.py
# Time complexity: O(n^2)
# Space complexity: O(1)

# Brute Force Solution
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        for i in range(len(nums)): # O(n)
            for j in range(len(nums)): # O(n)
                # when the sum of different element equal to the target
                if i != j and nums[i]+nums[j] == target:
                    # return the index of both element
                    return [i,j]
                
# Time complexity: O(n)
# Space complexity: O(1)      
# Hash Table

class Solution(object):
    def twoSum(self, nums, target):
        val = {}

        for i in range(len(nums)):
            # find the complement
            x = target - nums[i]
            # if complement is in the table return the result
            if x in val.keys():
                return [val[x],i]
            # add the value and it's index to table
            val[nums[i]]= i

File: 1-Two Sum.cpp
#include <iostream>
#include <vector>

// Time complexity: O(n^2)
// Space complexity: O(1)      
// bruce force

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size();i++){
            for (int j = 0; j < nums.size();j++) {
                if (i != j && nums.at(i) + nums.at(j) == target) {
                    return vector<int>{i,j};
                }
            }
        }
        return vector<int>{};
    }
};

// Time complexity: O(n)
// Space complexity: O(1)   
// hash table

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> numMap;
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            if (numMap.count(complement)) {
                return {numMap[complement], i};
            }
            numMap[nums[i]] = i;
        }
        return {}; // No solution found
    }
};